### PR TITLE
Service Discovery: fix shutdown blocking indefinitely in some cases

### PR DESCRIPTION
**Pull request by ReubenBond**
The logic used to determine when it was safe to dispose was not triggering because we added an additional flag which was not covered by the count mask.

### CODE REVIEW

File: src/Microsoft.Extensions.ServiceDiscovery/Http/HttpServiceEndPointResolver.cs [hunk 1]
```diff
@@ -159,7 +159,7 @@ private sealed class ResolverEntry : IAsyncDisposable
     {
         private readonly ServiceEndPointResolver _resolver;
         private readonly IServiceEndPointSelector _selector;
-        private const ulong CountMask = unchecked((ulong)-1);
+        private const ulong CountMask = ~(RecentUseFlag | DisposingFlag);   
         private const ulong RecentUseFlag = 1UL << 61;
         private const ulong DisposingFlag = 1UL << 62;
```
Comment (eerhardt) [inline src/Microsoft.Extensions.ServiceDiscovery/Http/HttpServiceEndPointResolver.cs:164]: Is there a reason we are leaving the top bit unused? Why aren't these:

```C#
        private const ulong RecentUseFlag = 1UL << 62;
        private const ulong DisposingFlag = 1UL << 63;
```

File: src/Microsoft.Extensions.ServiceDiscovery/Http/HttpServiceEndPointResolver.cs [hunk 2]
```diff
@@ -159,7 +159,7 @@ private sealed class ResolverEntry : IAsyncDisposable
     {
         private readonly ServiceEndPointResolver _resolver;
         private readonly IServiceEndPointSelector _selector;
-        private const ulong CountMask = unchecked((ulong)-1);
+        private const ulong CountMask = ~(RecentUseFlag | DisposingFlag);
```
Comment (eerhardt) [inline src/Microsoft.Extensions.ServiceDiscovery/Http/HttpServiceEndPointResolver.cs:162]:
```suggestion
        private const ulong CountMask = ~(RecentUseFlag | DisposingFlag);
```

**Review by eerhardt — APPROVED** [review]
Thanks!

File: src/Microsoft.Extensions.ServiceDiscovery/Http/HttpServiceEndPointResolver.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Microsoft.Extensions.ServiceDiscovery/Http/HttpServiceEndPointResolver.cs]: Maybe there was at some point, but not any more

**Review by eerhardt — APPROVED** [review]

**Comment by eerhardt** [comment]
@ReubenBond - did the build and tests run on this before you merged? It doesn't look like it to me.

**Pull request merged**